### PR TITLE
refactor(analysis): remove `_rewrite_filter()` in favor of using replacement patterns

### PR DESCRIPTION
**Pull request by kszucs**
`_rewrite_filter()` was mostly about the traversal itself, now it can be expressed with a simple pattern:

```py
node.replace(
    p.Reduction >> (lambda _: ops.TableArrayView(_.to_expr().as_table()), 
    filter=p.Value & ~p.WindowFunction & ~p.TableArrayView & ~p.ExistsSubquery
)
```

### CODE REVIEW

File: ibis/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql [hunk 1]
```diff
@@ -10,6 +10,6 @@ t1 AS (
 SELECT t1.*
 FROM t1
 WHERE t1.`a` = (
-  SELECT max(t1.`a`) AS `blah`
+  SELECT max(t1.`a`) AS `Max(a)`
```
Comment (kszucs) [inline ibis/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql:13]: The `Alias` is no longer pushed down into the subquery, but that should not be necessary.

File: ibis/expr/types/relations.py [hunk 2]
```diff
@@ -4176,15 +4173,17 @@ def _resolve_predicates(
...
+    rules = (
+        # turn reductions into table array views so that they can be used as
+        # WHERE t1.`a` = (SELECT max(t1.`a`) AS `Max(a)`
+        p.Reduction >> (lambda _: ops.TableArrayView(_.to_expr().as_table()))
+        |
+        # resolve unresolved exists subqueries to IN subqueries
+        p.UnresolvedExistsSubquery >> (lambda _: _.resolve(table.op()))
+    )
+    # do not apply the rules below the following nodes
+    until = p.Value & ~p.WindowFunction & ~p.TableArrayView & ~p.ExistsSubquery
```
Comment (kszucs) [inline ibis/expr/types/relations.py:4185]: Here we either could use a common base class or trait to check against, or rethink how we treat the current `Reduction` nodes. Actually reductions are not *valid* on their own, we usually rewrite them to aggregations, subqueries or window functions. 

Perhaps we should treat them as API-level constructs? That way rewrites could get simpler, but we should still fit them somehow into the `Value` node hierarchy.

Comment (cpcloud) [reply ibis/expr/types/relations.py:4185]: Why are reductions not valid on their own?

Isn't `t.x.sum()` a perfectly valid query?

File: ibis/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql (hunk 1 above)
Comment (cpcloud) [reply ibis/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql:13]: Yep, the alias isn't needed here.

File: ibis/tests/sql/snapshots/test_compiler/test_subquery_where_location/decompiled.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 param = ibis.param("timestamp")
 proj = alltypes.select(
     [alltypes.float_col, alltypes.timestamp_col, alltypes.int_col, alltypes.string_col]
-).filter(alltypes.timestamp_col < param)
+).filter(alltypes.timestamp_col < param.name("my_param"))
```
Comment (cpcloud) [inline ibis/tests/sql/snapshots/test_compiler/test_subquery_where_location/decompiled.py:16]: Is this change necessary?

Comment (cpcloud) [reply ibis/tests/sql/snapshots/test_compiler/test_subquery_where_location/decompiled.py:16]: Oh, I guess it's a generated code change, nevermind!

**Review by cpcloud — APPROVED** [review]
Sweet!

File: ibis/expr/types/relations.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:4185]: Ah, got it! Yeah, there's no valid SQL that corresponds to only the aggregation function call: you need a `SELECT`.
